File: aizu online judge/練習/3章 初等的整列/stable_sort.py
"""
https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_2_C&lang=jp
"""

def bubbleSort(a, n):
    flag = True
    i = 0
    while flag:
        flag = False
        for j in range(n - 1, i, -1):
            if a[j][1] < a[j - 1][1]:
                a[j], a[j - 1] = a[j - 1], a[j]
                flag = True
        i += 1
    print(" ".join(list(map(str, a))))

def selectionSort(a, n):
    for i in range(n):
        minj = i
        for j in range(i, n):
            if a[j][1] < a[minj][1]:
                minj = j
        a[i], a[minj] = a[minj], a[i]
    print(" ".join(list(map(str, a))))



n = int(input())
a = list(map(str, input().split()))
b = a[:]
bubbleSort(a, n)
print("Stable")
selectionSort(b, n)
if a == b:
    print("Stable")
else:
    print("Not stable")

File: aizu online judge/練習/3章 初等的整列/C++/stable_sort.cpp
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_2_C&lang=jp

#include <iostream>
using namespace std;

struct Card { char suit, value; };

void bubbleSort(struct Card A[], int N) {
  for (int i = 0; i < N; i++) {
    for (int j = N - 1; j >= i + 1; j--) {
      if (A[j].value < A[j-1].value) {
        swap(A[j], A[j-1]);
      };
    }
  }
}

void selectionSort(struct Card A[], int N) {
  for (int i = 0; i < N; i++) {
    int minj = i;
    for (int j = i; j < N; j++) {
      if (A[j].value < A[minj].value) {
        minj = j;
      }
    }
    swap(A[i], A[minj]);
  }
}

bool isStable(struct Card C1[], struct Card C2[], int N) {
  for (int i = 0; i < N; i++) {
    if (C1[i].suit != C2[i].suit) {
      return false;
    }
  }
  return true;
}

void print(struct Card A[], int N) {
  for (int i = 0; i < N; i++) {
    if (i) {
      cout << " ";
    }
    cout << A[i].suit << A[i].value;
  }
  cout << endl;
}

int main() {
  Card C1[100], C2[100];
  int N;
  
  cin >> N;
  for (int i = 0; i < N; i++) {
    cin >> C1[i].suit >> C1[i].value;
  }

  for (int i = 0; i < N; i++) {
    C2[i] = C1[i];
  }

  bubbleSort(C1, N);
  selectionSort(C2, N);

  print(C1, N);
  cout << "Stable" << endl;

  print(C2, N);
  if (isStable(C1, C2, N)) {
    cout << "Stable" << endl;
  } else {
    cout << "Not stable" << endl;
  }

  return 0;
}